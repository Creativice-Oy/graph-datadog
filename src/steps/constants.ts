import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const ACCOUNT_ENTITY_DATA_KEY = 'entity:account';

export const Steps = {
  USERS: 'fetch-users',
  ACCOUNT: 'fetch-account',
  ROLES: 'fetch-roles',
  BUILD_ACCOUNT_USER_RELATIONSHIPS: 'build-account-user-relationships',
  BUILD_ACCOUNT_ROLE_RELATIONSHIPS: 'build-account-role-relationships',
  BUILD_USER_ROLE_RELATIONSHIPS: 'build-user-role-relationships',
};

export const Entities: Record<'USER' | 'ACCOUNT' | 'ROLE', StepEntityMetadata> =
  {
    USER: {
      resourceName: 'User',
      _type: 'datadog_user',
      _class: ['User'],
    },
    ACCOUNT: {
      resourceName: 'Account',
      _type: 'datadog_account',
      _class: ['User'],
    },
    ROLE: {
      resourceName: 'Role',
      _type: 'datadog_role',
      _class: ['AccessRole'],
    },
  };

export const Relationships: Record<
  'ACCOUNT_HAS_USER' | 'ACCOUNT_HAS_ROLE' | 'USER_ASSIGNED_ROLE',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_USER: {
    _type: 'datadog_account_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.USER._type,
  },
  ACCOUNT_HAS_ROLE: {
    _type: 'datadog_account_has_role',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.ROLE._type,
  },
  USER_ASSIGNED_ROLE: {
    _type: 'datadog_user_assigned_role',
    _class: RelationshipClass.ASSIGNED,
    sourceType: Entities.USER._type,
    targetType: Entities.ROLE._type,
  },
};
