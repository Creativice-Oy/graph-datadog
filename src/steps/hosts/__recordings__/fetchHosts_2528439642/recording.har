{
  "log": {
    "_recordingName": "fetchHosts",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "77cabb0c6de3582c95984a1a86d0b0ae",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "datadog-api-client-typescript/1.2.0 (node 16.17.0; os Linux; arch x64)"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "dd-api-key",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "dd-application-key",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.datadoghq.com"
            }
          ],
          "headersSize": 348,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.datadoghq.com/api/v1/hosts"
        },
        "response": {
          "bodySize": 1668,
          "content": {
            "mimeType": "application/json",
            "size": 1668,
            "text": "{\"host_list\":[{\"id\":9231106664,\"tags_by_source\":{\"Datadog\":[\"host:mindfield\"]},\"aliases\":[\"mindfield\"],\"apps\":[\"agent\",\"container\",\"docker\",\"ntp\"],\"sources\":[\"agent\"],\"name\":\"mindfield\",\"host_name\":\"mindfield\",\"up\":true,\"last_reported_time\":1670945215,\"is_muted\":false,\"mute_timeout\":null,\"meta\":{\"processor\":\"AMD Ryzen 7 2700X Eight-Core Processor\",\"macV\":[\"\",\"\",\"\"],\"agent_flavor\":\"agent\",\"winV\":[\"\",\"\",\"\"],\"agent_version\":\"7.40.1\",\"fbsdV\":[\"\",\"\",\"\"],\"gohai\":\"{\\\"cpu\\\":{\\\"cache_size\\\":\\\"512 KB\\\",\\\"cpu_cores\\\":\\\"8\\\",\\\"cpu_logical_processors\\\":\\\"16\\\",\\\"family\\\":\\\"23\\\",\\\"mhz\\\":\\\"2200.000\\\",\\\"model\\\":\\\"8\\\",\\\"model_name\\\":\\\"AMD Ryzen 7 2700X Eight-Core Processor\\\",\\\"stepping\\\":\\\"2\\\",\\\"vendor_id\\\":\\\"AuthenticAMD\\\"},\\\"filesystem\\\":[{\\\"kb_size\\\":\\\"950914724\\\",\\\"mounted_on\\\":\\\"/\\\",\\\"name\\\":\\\"overlay\\\"},{\\\"kb_size\\\":\\\"65536\\\",\\\"mounted_on\\\":\\\"/dev\\\",\\\"name\\\":\\\"tmpfs\\\"},{\\\"kb_size\\\":\\\"65536\\\",\\\"mounted_on\\\":\\\"/dev/shm\\\",\\\"name\\\":\\\"shm\\\"},{\\\"kb_size\\\":\\\"950914724\\\",\\\"mounted_on\\\":\\\"/etc/hosts\\\",\\\"name\\\":\\\"/dev/nvme1n1p2\\\"},{\\\"kb_size\\\":\\\"1630628\\\",\\\"mounted_on\\\":\\\"/var/run/docker.sock\\\",\\\"name\\\":\\\"tmpfs\\\"},{\\\"kb_size\\\":\\\"8153124\\\",\\\"mounted_on\\\":\\\"/proc/asound\\\",\\\"name\\\":\\\"tmpfs\\\"},{\\\"kb_size\\\":\\\"8153124\\\",\\\"mounted_on\\\":\\\"/proc/acpi\\\",\\\"name\\\":\\\"tmpfs\\\"},{\\\"kb_size\\\":\\\"8153124\\\",\\\"mounted_on\\\":\\\"/proc/scsi\\\",\\\"name\\\":\\\"tmpfs\\\"},{\\\"kb_size\\\":\\\"8153124\\\",\\\"mounted_on\\\":\\\"/sys/firmware\\\",\\\"name\\\":\\\"tmpfs\\\"}],\\\"memory\\\":{\\\"swap_total\\\":\\\"9229308kB\\\",\\\"total\\\":\\\"16306252kB\\\"},\\\"network\\\":null,\\\"platform\\\":{\\\"GOOARCH\\\":\\\"amd64\\\",\\\"GOOS\\\":\\\"linux\\\",\\\"goV\\\":\\\"1.18.8\\\",\\\"hardware_platform\\\":\\\"x86_64\\\",\\\"hostname\\\":\\\"f992d058a188\\\",\\\"kernel_name\\\":\\\"Linux\\\",\\\"kernel_release\\\":\\\"5.15.0-56-generic\\\",\\\"kernel_version\\\":\\\"#62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022\\\",\\\"machine\\\":\\\"x86_64\\\",\\\"os\\\":\\\"GNU/Linux\\\",\\\"processor\\\":\\\"x86_64\\\",\\\"pythonV\\\":\\\"3.8.14\\\"}}\",\"network\":null,\"agent_checks\":[[\"ntp\",\"ntp\",\"ntp:d884b5186b651429\",\"OK\",\"\",\"\"],[\"memory\",\"memory\",\"memory\",\"OK\",\"\",\"\"],[\"load\",\"load\",\"load\",\"OK\",\"\",\"\"],[\"file_handle\",\"file_handle\",\"file_handle\",\"OK\",\"\",\"\"],[\"uptime\",\"uptime\",\"uptime\",\"OK\",\"\",\"\"],[\"network\",\"network\",\"network:d884b5186b651429\",\"OK\",\"\",\"\"],[\"disk\",\"disk\",\"disk:e5dffb8bef24336f\",\"OK\",\"\",\"\"],[\"io\",\"io\",\"io\",\"OK\",\"\",\"\"],[\"cpu\",\"cpu\",\"cpu\",\"OK\",\"\",\"\"],[\"container\",\"container\",\"container\",\"OK\",\"\",\"\"],[\"docker\",\"docker\",\"docker\",\"OK\",\"\",\"\"]],\"cpuCores\":1,\"logs_agent\":{\"auto_multi_line_detection_enabled\":false,\"transport\":\"\"},\"pythonV\":\"3.8.14\",\"platform\":\"linux\",\"machine\":\"amd64\",\"timezones\":[\"UTC\"],\"socket-hostname\":\"f992d058a188\",\"socket-fqdn\":\"f992d058a188\",\"container_meta\":{\"docker_swarm\":\"inactive\",\"docker_version\":\"20.10.21\"},\"nixV\":[\"ubuntu\",\"22.04\",\"\"],\"install_method\":{\"tool\":\"docker\",\"installer_version\":\"docker\",\"tool_version\":\"docker\"},\"host_id\":9231106664},\"metrics\":{\"cpu\":9.317546,\"iowait\":0.0037762218,\"load\":0.09271307}}],\"total_returned\":1,\"total_matching\":1,\"exact_total_matching\":true}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Tue, 13 Dec 2022 15:39:33 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=15724800;"
            },
            {
              "name": "content-security-policy",
              "value": "[REDACTED]"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            }
          ],
          "headersSize": 521,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-12-13T15:39:32.974Z",
        "time": 650,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 650
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
