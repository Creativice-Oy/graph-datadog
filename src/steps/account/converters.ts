import { User } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v1/models/User';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

function getAccountKey(id: string): string {
  return `datadog_account:${id}`;
}

export function createAccountEntity(account: User) {
  return createIntegrationEntity({
    entityData: {
      source: account,
      assign: {
        _class: Entities.ACCOUNT._class,
        _type: Entities.ACCOUNT._type,
        _key: getAccountKey(account.email?.toString() as string),
        name: account.email,
        username: account.handle,
        email: account.email,
        id: account.email,
        verified: account.verified,
        disabled: account.disabled,
        icon: account.icon,
        displayName: account.name || '',
        accessRole: account.accessRole,
      },
    },
  });
}
