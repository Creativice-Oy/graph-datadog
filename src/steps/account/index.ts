import {
  IntegrationStep,
  IntegrationStepExecutionContext,
  IntegrationWarnEventName,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../../client';
import { IntegrationConfig } from '../../config';
import { ACCOUNT_ENTITY_DATA_KEY, Entities, Steps } from '../constants';
import {
  createAccountEntity,
  createOrganizationAccountEntity,
} from './converters';

export async function fetchAccount({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  if (instance.config.datadogOrgPublicId) {
    let accountEntity;

    try {
      const organization = await apiClient.fetchOrganization(
        instance.config.datadogOrgPublicId,
      );

      accountEntity = organization.org
        ? createOrganizationAccountEntity(organization.org)
        : createAccountEntity(instance.name);
    } catch (err) {
      logger.warn({ err }, 'Unable to fetch Datadog organization information');
      logger.publishWarnEvent({
        name: IntegrationWarnEventName.MissingPermission,
        description:
          'Unable to fetch Datadog organization information. Please validate your Datadog organization public ID integration configuration value.',
      });

      accountEntity = createAccountEntity(instance.name);
    }

    await jobState.addEntity(accountEntity);
    await jobState.setData(ACCOUNT_ENTITY_DATA_KEY, accountEntity);
  } else {
    const accountEntity = await jobState.addEntity(
      createAccountEntity(instance.name),
    );

    await jobState.setData(ACCOUNT_ENTITY_DATA_KEY, accountEntity);

    logger.info('Customer did not supply organizationPublicId in config');
    logger.publishWarnEvent({
      name: IntegrationWarnEventName.MissingPermission,
      description:
        'Configure the organization public ID integration field to obtain additional information on datadog_account',
    });
  }
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.ACCOUNT,
    name: 'Fetch Account',
    entities: [Entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccount,
  },
];
