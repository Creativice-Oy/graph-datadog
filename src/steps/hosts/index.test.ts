import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { fetchHosts } from '.';
import { integrationConfig } from '../../../test/config';
import { setupDatadogRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../config';
import { fetchAccount } from '../account';
import { Entities, Relationships } from '../constants';

describe('#fetchHosts', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupDatadogRecording({
      directory: __dirname,
      name: 'fetchHosts',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAccount(context);
    await fetchHosts(context);

    const hosts = context.jobState.collectedEntities.filter((e) =>
      e._type.includes(Entities.HOST._type),
    );

    const accounts = context.jobState.collectedEntities.filter((e) =>
      e._type.includes(Entities.ACCOUNT._type),
    );

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(hosts.length).toBeGreaterThan(0);
    console.debug(JSON.stringify(hosts));
    expect(hosts).toMatchGraphObjectSchema({
      _class: Entities.HOST._class,
      schema: {
        additionalProperties: false,
        required: [],
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: Entities.HOST._type },
          _class: { const: Entities.HOST._class },
          _key: { type: 'string' },
          id: { type: 'string' },
          aliases: {
            type: 'array',
            items: { type: 'string' },
          },
          apps: {
            type: 'array',
            items: { type: 'string' },
          },
          awsName: { type: 'string' },
          hostname: { type: 'string' },
          isMuted: { type: 'boolean' },
          lastReportedTime: { type: 'number' },
          name: { type: 'string' },
          sources: {
            type: 'array',
            items: { type: 'string' },
          },
          icon: {
            type: 'array',
            items: { type: 'string' },
          },
          state: { type: 'string' },
        },
      },
    });

    expect(accounts.length).toEqual(1);
    expect(accounts).toMatchGraphObjectSchema({
      _class: Entities.ACCOUNT._class,
      schema: {
        additionalProperties: false,
        required: [],
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: Entities.ACCOUNT._type },
          _class: { const: Entities.ACCOUNT._class },
          _key: { type: 'string' },
          username: { type: 'string' },
          email: { type: 'string' },
          id: { type: 'string' },
          name: { type: 'string' },
          verified: { type: 'boolean' },
          disabled: { type: 'boolean' },
          icon: { type: 'string' },
          displayName: { type: 'string' },
          accessRole: { type: 'string' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === Relationships.ACCOUNT_HAS_HOST._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: { const: 'datadog_account_has_host' },
        },
      },
    });
  });
});
