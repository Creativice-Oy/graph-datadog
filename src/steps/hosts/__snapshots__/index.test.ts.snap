// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchHosts should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "datadog_account:viragsf@gmail.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": User {
            "accessRole": "adm",
            "disabled": false,
            "email": "viragsf@gmail.com",
            "handle": "viragsf@gmail.com",
            "icon": "https://secure.gravatar.com/avatar/edd1469180957cee25d7753ef4a84a13?s=48&d=retro",
            "name": null,
            "verified": true,
          },
        },
      ],
      "_type": "datadog_account",
      "accessRole": "adm",
      "createdOn": undefined,
      "disabled": false,
      "displayName": "viragsf@gmail.com",
      "email": "viragsf@gmail.com",
      "icon": "https://secure.gravatar.com/avatar/edd1469180957cee25d7753ef4a84a13?s=48&d=retro",
      "id": "viragsf@gmail.com",
      "name": "viragsf@gmail.com",
      "username": "viragsf@gmail.com",
      "verified": true,
    },
    Object {
      "_class": Array [
        "Host",
      ],
      "_key": "datadog_host:6264691456",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Host {
            "aliases": Array [
              "Dhaiwats-MacBook-Air.local",
            ],
            "apps": Array [
              "ntp",
              "agent",
            ],
            "awsName": undefined,
            "hostName": "Dhaiwats-MacBook-Air.local",
            "id": 6264691456,
            "isMuted": false,
            "lastReportedTime": 1660563750,
            "meta": HostMeta {
              "nixV": Array [
                null,
                null,
                null,
              ],
            },
            "metrics": HostMetrics {
              "cpu": null,
              "iowait": null,
              "load": null,
            },
            "muteTimeout": null,
            "name": "Dhaiwats-MacBook-Air.local",
            "sources": Array [
              "agent",
            ],
            "tagsBySource": Object {
              "Datadog": Array [
                "host:Dhaiwats-MacBook-Air.local",
              ],
            },
            "up": true,
          },
        },
      ],
      "_type": "datadog_host",
      "aliases": Array [
        "Dhaiwats-MacBook-Air.local",
      ],
      "apps": Array [
        "ntp",
        "agent",
      ],
      "awsName": undefined,
      "createdOn": undefined,
      "displayName": "Dhaiwats-MacBook-Air.local",
      "hostname": "Dhaiwats-MacBook-Air.local",
      "id": "6264691456",
      "isMuted": false,
      "lastReportedTime": 1660563750,
      "name": "Dhaiwats-MacBook-Air.local",
      "sources": Array [
        "agent",
      ],
      "state": "running",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "datadog_account:viragsf@gmail.com",
      "_key": "datadog_account:viragsf@gmail.com|has|datadog_host:6264691456",
      "_toEntityKey": "datadog_host:6264691456",
      "_type": "datadog_account_has_host",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "datadog_account",
    "datadog_host",
    "datadog_account_has_host",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;
